cmd/headscale/cli/nodes.go

```diff
@@ -73,103 +73,82 @@ var registerNodeCmd = &cobra.Command{
 	Run: func(cmd *cobra.Command, args []string) {
-		n, err := cmd.Flags().GetString("namespace")
+		output, _ := cmd.Flags().GetString("output")
+		namespace, err := cmd.Flags().GetString("namespace")
 		if err != nil {
-			log.Fatalf("Error getting namespace: %s", err)
+			ErrorOutput(err, fmt.Sprintf("Error getting namespace: %s", err), output)
+			return
 		}
-		o, _ := cmd.Flags().GetString("output")
 
-		h, err := getHeadscaleApp()
-		if err != nil {
-			log.Fatalf("Error initializing: %s", err)
-		}
-		machineIDStr, err := cmd.Flags().GetString("key")
+		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		defer cancel()
+
+		client, conn := getHeadscaleGRPCClient(ctx)
+		defer conn.Close()
+
+		machineKey, err := cmd.Flags().GetString("key")
 		if err != nil {
-			log.Fatalf("Error getting machine ID: %s", err)
-		}
-		m, err := h.RegisterMachine(machineIDStr, n)
-		if strings.HasPrefix(o, "json") {
-			JsonOutput(m, err, o)
+			ErrorOutput(err, fmt.Sprintf("Error getting machine key from flag: %s", err), output)
 			return
 		}
+
+		request := &v1.RegisterMachineRequest{
+			Key:       machineKey,
+			Namespace: namespace,
+		}
+
+		response, err := client.RegisterMachine(ctx, request)
 		if err != nil {
-			fmt.Printf("Cannot register machine: %s\n", err)
+			ErrorOutput(err, fmt.Sprintf("Cannot register machine: %s\n", err), output)
 			return
 		}
-		fmt.Printf("Machine registered\n")
+
+		SuccessOutput(response.Machine, "Machine register", output)
 	},
 }
 
 var listNodesCmd = &cobra.Command{
 	Use:   "list",
 	Short: "List nodes",
 	Run: func(cmd *cobra.Command, args []string) {
-		n, err := cmd.Flags().GetString("namespace")
+		output, _ := cmd.Flags().GetString("output")
+		namespace, err := cmd.Flags().GetString("namespace")
 		if err != nil {
-			log.Fatalf("Error getting namespace: %s", err)
+			ErrorOutput(err, fmt.Sprintf("Error getting namespace: %s", err), output)
+			return
 		}
-		o, _ := cmd.Flags().GetString("output")
 
-		h, err := getHeadscaleApp()
-		if err != nil {
-			log.Fatalf("Error initializing: %s", err)
-		}
+		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		defer cancel()
 
-		var namespaces []headscale.Namespace
-		var namespace *headscale.Namespace
-		var sharedMachines *[]headscale.Machine
-		if len(n) == 0 {
-			// no namespace provided, list all
-			tmp, err := h.ListNamespaces()
-			if err != nil {
-				log.Fatalf("Error fetching namespace: %s", err)
-			}
-			namespaces = *tmp
-		} else {
-			namespace, err = h.GetNamespace(n)
-			if err != nil {
-				log.Fatalf("Error fetching namespace: %s", err)
-			}
-			namespaces = append(namespaces, *namespace)
+		client, conn := getHeadscaleGRPCClient(ctx)
+		defer conn.Close()
 
-			sharedMachines, err = h.ListSharedMachinesInNamespace(n)
-			if err != nil {
-				log.Fatalf("Error fetching shared machines: %s", err)
-			}
+		request := &v1.ListMachinesRequest{
+			Namespace: namespace,
```

Comment:
`namespace` is an empty string here in the default case, when the user does `headscale nodes list` without specifying a specific namespace.

This leads to a panic:
<pre>
Nov 06 20:24:42 panic: runtime error: invalid memory address or nil pointer dereference
Nov 06 20:24:42 [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xea707b]
Nov 06 20:24:42 goroutine 286 [running]:
Nov 06 20:24:42 github.com/juanfont/headscale.(*PreAuthKey).toProto(0xc0003939a0, 0xc0000d4780)
Nov 06 20:24:42         /usr/src/headscale-krdalby/headscale/preauth_keys.go:149 +0x5b
Nov 06 20:24:42 github.com/juanfont/headscale.(*Machine).toProto(0xc0006d96c8, 0xc0000d46c0)
Nov 06 20:24:42         /usr/src/headscale-krdalby/headscale/machine.go:544 +0x3f6
Nov 06 20:24:42 github.com/juanfont/headscale.headscaleV1APIServer.ListMachines(0xc0003f4580, 0x13d1318, 0xc00092ff50, 0xc000268ac0, 0x1208560, 0xc00018fa00, 0x13df7d0)
Nov 06 20:24:42         /usr/src/headscale-krdalby/headscale/grpcv1.go:228 +0x465
Nov 06 20:24:42 github.com/juanfont/headscale/gen/go/headscale/v1._HeadscaleService_ListMachines_Handler.func1(0x13d1318, 0xc00092ff50, 0x11514e0, 0xc000268ac0, 0x0, 0x203000, 0x7f67bc21b328, 0x18)
Nov 06 20:24:42         /usr/src/headscale-krdalby/headscale/gen/go/headscale/v1/headscale_grpc.pb.go:530 +0x89
Nov 06 20:24:42 github.com/philip-bui/grpc-zerolog.NewUnaryServerInterceptorWithLogger.func1(0x13d1318, 0xc00092ff50, 0x11514e0, 0xc000268ac0, 0xc00048dbe0, 0xc000751a58, 0xc000971ba0, 0x541266, 0x119ab00, 0xc00092ff50)
Nov 06 20:24:42         /usr/src/go/pkg/mod/github.com/philip-bui/grpc-zerolog@v1.0.1/unary_interceptor.go:50 +0x9d
Nov 06 20:24:42 github.com/juanfont/headscale/gen/go/headscale/v1._HeadscaleService_ListMachines_Handler(0x1208560, 0xc000011278, 0x13d1318, 0xc00092ff50, 0xc00074f200, 0xc000135740, 0x13d1318, 0xc00092ff50, 0x0, 0x0)
Nov 06 20:24:42         /usr/src/headscale-krdalby/headscale/gen/go/headscale/v1/headscale_grpc.pb.go:532 +0x150
Nov 06 20:24:42 google.golang.org/grpc.(*Server).processUnaryRPC(0xc00053bdc0, 0x13db258, 0xc000667980, 0xc000175320, 0xc000290d50, 0x19f1a40, 0x0, 0x0, 0x0)
Nov 06 20:24:42         /usr/src/go/pkg/mod/google.golang.org/grpc@v1.42.0/server.go:1282 +0x52b
Nov 06 20:24:42 google.golang.org/grpc.(*Server).handleStream(0xc00053bdc0, 0x13db258, 0xc000667980, 0xc000175320, 0x0)
Nov 06 20:24:42         /usr/src/go/pkg/mod/google.golang.org/grpc@v1.42.0/server.go:1616 +0xd0c
Nov 06 20:24:42 google.golang.org/grpc.(*Server).serveStreams.func1.2(0xc000036ec0, 0xc00053bdc0, 0x13db258, 0xc000667980, 0xc000175320)
Nov 06 20:24:42         /usr/src/go/pkg/mod/google.golang.org/grpc@v1.42.0/server.go:921 +0xab
Nov 06 20:24:42 created by google.golang.org/grpc.(*Server).serveStreams.func1
Nov 06 20:24:42         /usr/src/go/pkg/mod/google.golang.org/grpc@v1.42.0/server.go:919 +0x1fd
</pre>

Reply:
As a side note; when we panic the socket file is left behind; this means that headscale can't start anymore because it panics on startup with `panic: listen unix /var/run/headscale.sock: bind: address already in use`. Can we make it handle an existing socket file?

Reply:
Somehow I could not reproduce that error. Just started headscale `heascale serve` and then run `heascale namespaces list`. It returned:
```
ID | Name | NodeKey | Namespace | IP address | Ephemeral | Last seen | Online
```

Reply:
> `namespace` is an empty string here in the default case, when the user does `headscale nodes list` without specifying a specific namespace.
> 
> This leads to a panic:
> 
> Nov 06 20:24:42 panic: runtime error: invalid memory address or nil pointer dereference
> Nov 06 20:24:42 [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xea707b]
> Nov 06 20:24:42 goroutine 286 [running]:
> Nov 06 20:24:42 github.com/juanfont/headscale.(*PreAuthKey).toProto(0xc0003939a0, 0xc0000d4780)
> Nov 06 20:24:42         /usr/src/headscale-krdalby/headscale/preauth_keys.go:149 +0x5b
> Nov 06 20:24:42 github.com/juanfont/headscale.(*Machine).toProto(0xc0006d96c8, 0xc0000d46c0)
> Nov 06 20:24:42         /usr/src/headscale-krdalby/headscale/machine.go:544 +0x3f6
> Nov 06 20:24:42 github.com/juanfont/headscale.headscaleV1APIServer.ListMachines(0xc0003f4580, 0x13d1318, 0xc00092ff50, 0xc000268ac0, 0x1208560, 0xc00018fa00, 0x13df7d0)
> Nov 06 20:24:42         /usr/src/headscale-krdalby/headscale/grpcv1.go:228 +0x465
> Nov 06 20:24:42 github.com/juanfont/headscale/gen/go/headscale/v1._HeadscaleService_ListMachines_Handler.func1(0x13d1318, 0xc00092ff50, 0x11514e0, 0xc000268ac0, 0x0, 0x203000, 0x7f67bc21b328, 0x18)
> Nov 06 20:24:42         /usr/src/headscale-krdalby/headscale/gen/go/headscale/v1/headscale_grpc.pb.go:530 +0x89
> Nov 06 20:24:42 github.com/philip-bui/grpc-zerolog.NewUnaryServerInterceptorWithLogger.func1(0x13d1318, 0xc00092ff50, 0x11514e0, 0xc000268ac0, 0xc00048dbe0, 0xc000751a58, 0xc000971ba0, 0x541266, 0x119ab00, 0xc00092ff50)
> Nov 06 20:24:42         /usr/src/go/pkg/mod/github.com/philip-bui/grpc-zerolog@v1.0.1/unary_interceptor.go:50 +0x9d
> Nov 06 20:24:42 github.com/juanfont/headscale/gen/go/headscale/v1._HeadscaleService_ListMachines_Handler(0x1208560, 0xc000011278, 0x13d1318, 0xc00092ff50, 0xc00074f200, 0xc000135740, 0x13d1318, 0xc00092ff50, 0x0, 0x0)
> Nov 06 20:24:42         /usr/src/headscale-krdalby/headscale/gen/go/headscale/v1/headscale_grpc.pb.go:532 +0x150
> Nov 06 20:24:42 google.golang.org/grpc.(*Server).processUnaryRPC(0xc00053bdc0, 0x13db258, 0xc000667980, 0xc000175320, 0xc000290d50, 0x19f1a40, 0x0, 0x0, 0x0)
> Nov 06 20:24:42         /usr/src/go/pkg/mod/google.golang.org/grpc@v1.42.0/server.go:1282 +0x52b
> Nov 06 20:24:42 google.golang.org/grpc.(*Server).handleStream(0xc00053bdc0, 0x13db258, 0xc000667980, 0xc000175320, 0x0)
> Nov 06 20:24:42         /usr/src/go/pkg/mod/google.golang.org/grpc@v1.42.0/server.go:1616 +0xd0c
> Nov 06 20:24:42 google.golang.org/grpc.(*Server).serveStreams.func1.2(0xc000036ec0, 0xc00053bdc0, 0x13db258, 0xc000667980, 0xc000175320)
> Nov 06 20:24:42         /usr/src/go/pkg/mod/google.golang.org/grpc@v1.42.0/server.go:921 +0xab
> Nov 06 20:24:42 created by google.golang.org/grpc.(*Server).serveStreams.func1
> Nov 06 20:24:42         /usr/src/go/pkg/mod/google.golang.org/grpc@v1.42.0/server.go:919 +0x1fd

I am still seeing this with the latest code, do you want me to investigate @kradalby ?

Reply:
Ah, I should have posted an update, I didnt get to this one yet, will have a look after work, but if you have some time, yes please :)

Then we can add a test case to the cli test to verify and make sure we dont have regressions

Reply:
@cure Can you have a go now?, I think I might have fixed it